nntrainer/tensor/tensor.cpp

```diff
@@ -214,162 +216,184 @@ int Tensor::multiply_i(float const &value) {
-  if (value == 0.0f) {
-    throw std::runtime_error("Error: Divide by zero");
-  }
-
-  CLONE_OP_I(divide_i, value);
+  Tensor t;
+  return divide(value, t);
 }
 
-Tensor Tensor::divide(float const &value, Tensor &out) const {
-  throw nntrainer::exception::not_supported("NYI");
+Tensor &Tensor::divide(float const &value, Tensor &out) const {
+  auto f = std::bind(std::divides<float>(), std::placeholders::_1, value);
+  /// @todo add unittest
+  if (value == 0.0f) {
+    std::stringstream ss;
+    ss << "[Tensor] divide by value failed, value: " << value;
+    throw std::invalid_argument(ss.str().c_str());
+  }
+  return apply(f, out);
 }
 
 int Tensor::add_i(float const &value) {
-  float *data = getData();
-
-  std::transform(data, data + length(), data,
-                 std::bind2nd(std::plus<float>(), value));
-
+  this->add(value, *this);
   return ML_ERROR_NONE;
 }
 
-Tensor Tensor::add(float const &value) const { CLONE_OP_I(add_i, value); }
-
-Tensor Tensor::add(float const &value, Tensor &out) const {
-  const float *data = getData();
-
-  CREATE_IF_EMPTY_DIMS(out, getDim());
+Tensor Tensor::add(float const &value) const {
+  Tensor t;
+  return add(value, t);
+}
 
-  std::transform(data, data + length(), out.getData(),
-                 std::bind2nd(std::plus<float>(), value));
+Tensor &Tensor::add(float const &value, Tensor &out) const {
+  /// @todo add unittest
+  auto f = std::bind(std::plus<float>(), std::placeholders::_1, value);
+  return apply(f, out);
+}
 
-  return out;
+int Tensor::subtract_i(float const &value) {
+  this->subtract(value, *this);
+  return ML_ERROR_NONE;
 }
 
-int Tensor::subtract_i(float const &value) { return this->add_i(-value); }
+Tensor Tensor::subtract(float const &value) const {
+  Tensor t;
+  return subtract(value, t);
+}
 
-Tensor Tensor::subtract(float const &value) const { return this->add(-value); }
+Tensor &Tensor::subtract(float const &value, Tensor &out) const {
+  /// @todo add unittest
+  auto f = std::bind(std::minus<float>(), std::placeholders::_1, value);
+  return apply(f, out);
+}
 
-Tensor Tensor::subtract(float const &value, Tensor &out) const {
-  throw nntrainer::exception::not_supported("NYI");
+int Tensor::pow_i(float exponent) {
+  pow(exponent, *this);
+  return ML_ERROR_NONE;
 }
 
 Tensor Tensor::pow(float exponent) const {
-  return apply([=](float in) { return powf(in, exponent); });
+  Tensor t;
+  return pow(exponent, t);
 }
 
-int Tensor::pow_i(float exponent) {
-  return apply_i([=](float in) { return powf(in, exponent); });
+Tensor &Tensor::pow(float exponent, Tensor &out) const {
+  auto f = [exponent](float in) { return powf(in, exponent); };
+  return apply(f, out);
 }
 
 int Tensor::multiply_i(Tensor const &m) {
-  auto f = [&](const BroadcastInfo &e, float *buf, const float *m_buf) {
-    for (unsigned int i = 0; i < e.buffer_size; ++i) {
-      *buf *= *m_buf;
-      buf += strides[3];
-      m_buf += e.strides[3];
-    }
-  };
+  try {
```

Comment:
do we need to have `try catch` here?

Reply:
If this is because of the signature, update the signature to be void, but lets not add `try catch` in critical path.

Reply:
Also check this at other places as well.

Reply:
This is not newly added, so all the benchmarks was based on this data path.
As far as I know, overhead is significant only when the error occurs.

I think it is fine because
1. Throwing error is not at the critical path
1. Operation related error is supposed to return integer code in this semantic.

We could check if there is a noticeable difference with `noexcept` or not catching but I think this could be done later.

Reply:
Added issue #887

Reply:
don't we need catch the return value for c API?

Reply:
There is a throw boundary in capi so in capi side it's okay.